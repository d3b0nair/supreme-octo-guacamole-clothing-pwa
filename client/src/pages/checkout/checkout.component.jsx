import React, { lazy, Suspense, useState } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import CheckoutItem from "../../components/checkout-item/checkout-item.component";
import CustomButton from "../../components/custom-button/custom-button.component";
import Spinner from "../../components/spinner/spinner.component";
import PopUpMsg from "../../components/pop-up-msg/pop-up-msg.component";

import {
  selectCartItems,
  selectCartTotal,
} from "../../redux/cart/cart.selector";
import { selectCheckoutStatus } from "../../redux/checkout/checkout.selectors";
import { checkoutStart } from "../../redux/checkout/checkout.actions";
import { selectCurrentUser } from "../../redux/user/user.selector";

import {
  CheckoutPageContainer,
  TotalPrice,
  CartHeader,
  Cart,
  StripeContainer,
  CartContainer,
} from "./checkout.styles.jsx";

import history from "../../history";

const StripePayments = lazy(() =>
  import("../../components/stripe-payments/stripe-payments.component")
);

export function CheckoutPage({
  cartItems,
  total,
  checkoutStatus,
  checkoutStart,
  currentUser,
}) {
  const [showMsg, setShowMsg] = useState(false);
  const proceedToCheckout = () =>
    currentUser
      ? cartItems.length > 0
        ? checkoutStart()
        : null
      : setShowMsg(true);
  const StoreName = () => <h2>Supreme Octo Guacamole Cloting Ltd.</h2>;
  const CartRows = () => (
    <>
      <div>
        <span>Product</span>
      </div>
      <div>
        <span>Description</span>
      </div>
      <div>
        <span>Quantity</span>
      </div>
      <div>
        <span>Price</span>
      </div>
      <div>
        <span>Remove</span>
      </div>
    </>
  );
  const renderStoreName = checkoutStatus !== false ? <StoreName /> : null;
  const renderCartItems = cartItems.map((cartItem) => (
    <CheckoutItem key={cartItem.id} cartItem={cartItem} />
  ));
  const renderCheckoutWithStripe =
    checkoutStatus !== false ? (
      <StripeContainer>
        <Suspense fallback={<Spinner />}>
          <StripePayments />
        </Suspense>
      </StripeContainer>
    ) : null;
  const renderButtonToPromtCheckoutWithStripe =
    checkoutStatus === false ? (
      <>
        <TotalPrice>TOTAL: ${total}</TotalPrice>
        <CustomButton
          style={{ margin: "3vh 0 3vh 0" }}
          onClick={() => proceedToCheckout()}
        >
          Proceed with payment
        </CustomButton>
      </>
    ) : null;
  const renderPopUpMsg = showMsg ? (
    <PopUpMsg
      onDismiss={() => setShowMsg(false)}
      msg={"Please log in or create an account to checkout."}
      action={() => history.push("/signin")}
      actionText={"Do it"}
    />
  ) : null;

  return (
    <CheckoutPageContainer>
      {renderStoreName}
      <CartContainer>
        <Cart>
          <CartHeader>
            <CartRows />
          </CartHeader>
          {renderCartItems}
        </Cart>
        {renderCheckoutWithStripe}
      </CartContainer>
      {renderButtonToPromtCheckoutWithStripe}
      {renderPopUpMsg}
    </CheckoutPageContainer>
  );
}

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
  total: selectCartTotal,
  checkoutStatus: selectCheckoutStatus,
  currentUser: selectCurrentUser,
});

const mapDispatchToProps = (dispatch) => ({
  checkoutStart: () => dispatch(checkoutStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(CheckoutPage);
