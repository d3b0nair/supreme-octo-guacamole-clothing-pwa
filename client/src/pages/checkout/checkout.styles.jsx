import styled, { css } from "styled-components";

import { frostedGlass } from "./../../global.styles";

const RowSize = css`
  width: 100%;
  min-width: 320px;
  @media screen and (max-width: 800px) {
    min-width: auto;
  }
`;

export const container_media_margin = 10;
export const container_top_bottom_margin = 30;

export const CheckoutPageContainer = styled.div`
  width: 700px;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  margin: ${container_top_bottom_margin}px auto;
  h2 {
    text-align: center;
  }
  @media screen and (max-width: 800px) {
    margin: ${container_media_margin}px;
    width: auto;
  }
`;
export const CartHeader = styled.div`
  padding: 0 0 10px 0;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr;
  gap: 0px 0px;
  grid-template-areas: ". . . . .";
  border-bottom: 1px solid #ffffff;
  text-align: center;
  div {
    text-transform: capitalize;
  }
  @media screen and (max-width: 300px) {
    display: none;
  }
`;

export const TotalPrice = styled.div`
  margin-top: 30px;
  margin-left: auto;
  font-size: 36px;
  @media screen and (max-width: 800px) {
    font-size: 30px;
    margin-top: 20px;
    margin-right: auto;
  }
`;

export const CartContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: auto;
  width: 100%;
  min-width: 700px;
  @media screen and (max-width: 800px) {
    min-width: 85px;
    flex-direction: column;
  }
`;
export const Cart = styled.div`
  ${frostedGlass}
  padding: 10px;
  ${RowSize}
`;
export const StripeContainer = styled.div`
  margin: 0 25px 0 25px;

  @media screen and (max-width: 800px) {
    margin: 20px auto;
  }
  ${RowSize}
`;
