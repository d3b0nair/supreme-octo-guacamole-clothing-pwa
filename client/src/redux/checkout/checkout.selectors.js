import { createSelector } from "reselect";

const selectCheckout = (state) => state.checkout;
export const selectCheckoutStatus = createSelector(
  [selectCheckout],
  (checkout) => checkout.checkout_status
);
