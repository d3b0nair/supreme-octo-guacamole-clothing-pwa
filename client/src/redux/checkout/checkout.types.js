export const CheckoutActionTypes = {
  CHECKOUT_START: "CHECKOUT_START",
  CHECKOUT_PROCESSING: "CHECKOUT_PROCESSING",
  CHECKOUT_SUCCEDED: "CHECKOUT_SUCCEDED",
  CHECKOUT_RESET: "CHECKOUT_RESET",
  CREATE_ORDER_START: "CREATE_ORDER_START",
  CREATE_ORDER_FAIL: "CREATE_ORDER_FAIL",
  CREATE_ORDER_SUCCESS: "CREATE_ORDER_SUCCESS"
};
