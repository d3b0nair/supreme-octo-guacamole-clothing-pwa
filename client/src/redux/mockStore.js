import { createStore } from "redux";
import { persistReducer } from "redux-persist";
import { rootReducer, persistConfig } from "./root-reducer";
export const createMockStore = ({ mockState, mockReducers }) => {
  const combinedReducers = persistReducer(
    persistConfig,
    rootReducer(mockReducers)
  );
  return createStore(combinedReducers, mockState);
};
