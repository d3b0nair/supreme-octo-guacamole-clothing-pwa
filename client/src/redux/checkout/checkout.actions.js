import { CheckoutActionTypes } from "./checkout.types";

export const checkoutStart = () => ({
  type: CheckoutActionTypes.CHECKOUT_START,
});

export const processCheckout = () => ({
  type: CheckoutActionTypes.CHECKOUT_PROCESSING,
});

export const checkoutSucceded = () => ({
  type: CheckoutActionTypes.CHECKOUT_SUCCEDED,
});

export const checkoutReset = () => ({
  type: CheckoutActionTypes.CHECKOUT_RESET,
});

export const createOrder_START = (id, items, customer, total) => ({
  type: CheckoutActionTypes.CREATE_ORDER_START,
  payload: { id, items, customer, total },
});
export const createOrder_FAIL = () => ({
  type: CheckoutActionTypes.CREATE_ORDER_FAIL,
});
export const createOrder_SUCCESS = () => ({
  type: CheckoutActionTypes.CREATE_ORDER_SUCCESS,
});
