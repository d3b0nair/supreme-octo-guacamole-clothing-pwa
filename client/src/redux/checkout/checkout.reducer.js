import { CheckoutActionTypes } from "./checkout.types";

export const INITIAL_STATE = {
  checkout_status: false,
  checkout_error: null,
};

const checkoutReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CheckoutActionTypes.CHECKOUT_PROCESSING:
      return { ...state, checkout_status: "PROCESSING" };
    case CheckoutActionTypes.CHECKOUT_START:
      return { ...state, checkout_status: "START" };
    case CheckoutActionTypes.CHECKOUT_SUCCEDED:
      return { ...state, checkout_status: "SUCCEDED" };
    case CheckoutActionTypes.CHECKOUT_RESET:
      return { ...state, checkout_status: false };
    case CheckoutActionTypes.CREATE_ORDER_FAIL:
      return {
        ...state,
        checkout_error: action.payload,
      };
    case CheckoutActionTypes.CREATE_ORDER_SUCCESS:
      return {
        ...state,
        checkout_error: null,
      };
    default:
      return state;
  }
};

export default checkoutReducer;
