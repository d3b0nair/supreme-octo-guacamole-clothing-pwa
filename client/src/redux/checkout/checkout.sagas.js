import { all, call, takeLatest, put } from "redux-saga/effects";

import { CheckoutActionTypes } from "./checkout.types";
import history from "../../history";

import { clearCart } from "../cart/cart.actions";
import {
  checkoutReset,
  createOrder_SUCCESS,
  checkoutSucceded,
  createOrder_FAIL,
} from "./checkout.actions";

export function* resetCheckoutAndCart() {
  yield put(checkoutSucceded());
  yield put(checkoutReset());
  yield put(clearCart());
}

export function* createOrderSTART({
  payload: { id, items, customer, total, userId },
}) {
  try {
    yield call(fetch, "/create-order", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id, items, customer, total, userId }),
    });
    yield put(createOrder_SUCCESS());
    yield call(resetCheckoutAndCart);
    yield call(history.push, `/invoice/${id}`);
  } catch (error) {
    console.log(error);
    yield put(createOrder_FAIL(error));
  }
}

export function* onCheckoutSuccess() {
  yield takeLatest(CheckoutActionTypes.CREATE_ORDER_START, createOrderSTART);
}

export function* checkoutSagas() {
  yield all([call(onCheckoutSuccess)]);
}
