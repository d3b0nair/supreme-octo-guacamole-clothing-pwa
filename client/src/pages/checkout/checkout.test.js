import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { createMockStore } from "../../redux/mockStore";
import { Provider } from "react-redux";
import { CheckoutPage } from "./checkout.component";
const mockReducer = (state = {}, action) => state;

const store = createMockStore({
  state: {},
  reducers: {
    checkout: mockReducer,
  },
});

const mockCheckoutStart = jest.fn();
const mockProps = (otherProps) => ({
  ...otherProps,
  total: 500,
  checkoutStatus: false,
  checkoutStart: mockCheckoutStart,
  currentUser: { displayName: "username" },
  cartItems: [{ id: 1 }, { id: 2 }],
});
const component = (mockProps) => (
  <Provider store={store}>
    <CheckoutPage {...mockProps} />
  </Provider>
);

afterEach(cleanup);

describe("Checkout component", () => {
  it("should mount", () => {
    const { container } = render(component(mockProps({ cartItems: [] })));
    expect(container).toMatchSnapshot();
  });
  it("if currentUser exsits and cartItems is not empty user can proceed with payment on a button click", async () => {
    render(component(mockProps()));
    await waitFor(() => fireEvent.click(screen.getByRole("button")));
    expect(mockCheckoutStart).toHaveBeenCalled();
  });
});
