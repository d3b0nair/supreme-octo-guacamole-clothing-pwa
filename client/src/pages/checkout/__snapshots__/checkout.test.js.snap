// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout component should mount 1`] = `
<div>
  <div
    class="sc-bqiRlB jFiSo"
  >
    <div
      class="sc-fotOHu iYUZQW"
    >
      <div
        class="sc-fFeiMQ hwlXzM"
      >
        <div
          class="sc-ksdxgE bbYwPF"
        >
          <div>
            <span>
              Product
            </span>
          </div>
          <div>
            <span>
              Description
            </span>
          </div>
          <div>
            <span>
              Quantity
            </span>
          </div>
          <div>
            <span>
              Price
            </span>
          </div>
          <div>
            <span>
              Remove
            </span>
          </div>
        </div>
        <div
          class="sc-gsDKAQ ksGHNo"
        >
          <div
            class="sc-dkPtRN gUwQJy"
          >
            <img
              alt="item"
              class="sc-hKwDye gkmiPh"
            />
          </div>
          <span
            class="sc-eCImPb fqRrfq"
          />
          <span
            class="sc-jRQBWg fGFNQO"
          >
            <div
              class="sc-gKclnd lcsmPG"
              data-testid="remove-checkout-item"
            >
              ❮
            </div>
            <div
              class="sc-iCfMLu jHWOzV"
            />
            <div
              class="sc-gKclnd lcsmPG"
              data-testid="add-checkout-item"
            >
              ❯
            </div>
          </span>
          <span
            class="sc-eCImPb fqRrfq"
          >
            $
          </span>
          <div
            class="sc-pVTFL fjkOjk"
          >
            <div
              class="sc-furwcr gxgspo"
              data-testid="clear-checkout-item"
            >
              ✕
            </div>
          </div>
        </div>
        <div
          class="sc-gsDKAQ ksGHNo"
        >
          <div
            class="sc-dkPtRN gUwQJy"
          >
            <img
              alt="item"
              class="sc-hKwDye gkmiPh"
            />
          </div>
          <span
            class="sc-eCImPb fqRrfq"
          />
          <span
            class="sc-jRQBWg fGFNQO"
          >
            <div
              class="sc-gKclnd lcsmPG"
              data-testid="remove-checkout-item"
            >
              ❮
            </div>
            <div
              class="sc-iCfMLu jHWOzV"
            />
            <div
              class="sc-gKclnd lcsmPG"
              data-testid="add-checkout-item"
            >
              ❯
            </div>
          </span>
          <span
            class="sc-eCImPb fqRrfq"
          >
            $
          </span>
          <div
            class="sc-pVTFL fjkOjk"
          >
            <div
              class="sc-furwcr gxgspo"
              data-testid="clear-checkout-item"
            >
              ✕
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-hBUSln gnsOKe"
    >
      TOTAL: $
      500
    </div>
    <button
      class="sc-bdvvtL fNaNEb"
      style="margin: 3vh 0px 3vh 0px;"
    >
      Proceed with payment
    </button>
  </div>
</div>
`;
